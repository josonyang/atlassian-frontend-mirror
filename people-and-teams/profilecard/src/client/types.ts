export type DEPRECATED_ErrorAttributes = {
	errorCount?: number;
	errorDetails?: Omit<DEPRECATED_ErrorAttributes, 'errorDetails'>[];
	errorMessage: string;
	errorCategory?: string;
	errorType?: string;
	errorPath?: string;
	errorNumber?: number;
	isSLOFailure: boolean;
	traceId?: string | null;
	errorStatusCode?: number;
};

export type ErrorAttributes = {
	errorCount: number | null;
	errorDetails: Omit<ErrorAttributes, 'errorDetails'>[] | null;
	errorMessage: string;
	errorCategory: string | null;
	errorType: string | null;
	errorPath: string | null;
	errorNumber: number | null;
	isSLOFailure: boolean;
	traceId: string | null;
	errorStatusCode: number | null;
};
