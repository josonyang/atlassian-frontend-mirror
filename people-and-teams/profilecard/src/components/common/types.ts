import { type AnalyticsEventPayload } from '@atlaskit/analytics-next';
import { type PopupProps } from '@atlaskit/popup';

import { type ProfileCardErrorType } from '../../types';

export type ProfileCardTriggerProps<T> = {
	trigger: 'hover' | 'click';
	ariaLabelledBy?: string;
	disabledAriaAttributes?: boolean;
	children: React.ReactNode;
	renderProfileCard: ({
		profileData,
		error,
	}: {
		profileData?: T;
		error: ProfileCardErrorType | undefined | null;
	}) => React.ReactNode;
	fetchProfile?: () => Promise<T>;
	profileCardType: 'agent' | 'user' | 'team';
	fireAnalytics?: (payload: AnalyticsEventPayload) => void;
} & Omit<PopupProps, 'trigger' | 'isOpen' | 'content'>;
