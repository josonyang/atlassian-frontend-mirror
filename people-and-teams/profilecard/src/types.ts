import type React from 'react';

import { type IntlShape } from 'react-intl-next';

import { type AnalyticsEventPayload, type CreateUIAnalyticsEvent } from '@atlaskit/analytics-next';
import { type ConversationStarter } from '@atlaskit/rovo-agent-components';

import { type TeamCentralScopes } from './client/ProfileCardClient';
import type RovoAgentCardClient from './client/RovoAgentCardClient';
import {
	type default as TeamCentralCardClient,
	type TeamCentralCardClientOptions,
} from './client/TeamCentralCardClient';
import type TeamProfileCardClient from './client/TeamProfileCardClient';
import type UserProfileCardClient from './client/UserProfileCardClient';

export interface ApiClientResponse {
	User: {
		id: string;
		accountType: string;
		isBot: boolean;
		isCurrentUser: boolean;
		avatarUrl: string | null;
		email: string | null;
		fullName: string | null;
		location: string | null;
		meta: string | null;
		nickname: string | null;
		companyName: string | null;
		remoteTimeString: string | null;
		remoteWeekdayIndex: string | null;
		remoteWeekdayString: string | null;
		status: StatusType;
		statusModifiedDate: number | null;
	};
}

type FlagAppearance = 'error' | 'info' | 'normal' | 'success' | 'warning';
export interface Flag {
	id: number | string;
	title?: string | React.ReactNode;
	description?: string | React.ReactNode;
	type?: FlagAppearance;

	appearance?: FlagAppearance;
	actions?: {
		content: React.ReactNode;
		onClick?: () => void;
		href?: string;
		target?: string;
	}[];
	icon?: JSX.Element;
}

export interface Team {
	// Header image props
	largeAvatarImageUrl?: string;
	smallAvatarImageUrl?: string;
	largeHeaderImageUrl?: string;
	smallHeaderImageUrl?: string;
	// Regular team details
	id: string;
	displayName: string;
	description: string;
	organizationId?: string;
	members?: {
		id: string;
		fullName: string;
		avatarUrl: string;
	}[];
	isVerified?: boolean;
}

export interface RovoAgent {
	id: string;
	identity_account_id?: string | null;
	named_id: string;
	name: string;
	description: string | null;
	system_prompt_template?: string | null;
	creator_type: 'SYSTEM' | 'CUSTOMER' | 'THIRD_PARTY' | 'FORGE';
	creator?: string | null;
	visibility?: 'PUBLIC' | 'PRIVATE' | null;
	is_default: boolean;
	actor_type: 'AGENT';
	follow_up_prompt_template?: string | null;
	plugin_routing_type?: 'DEFAULT' | 'SKIP' | null;
	user_defined_conversation_starters: string[] | null;
	favourite: boolean;
	favourite_count: number;
	deactivated?: boolean;
	icon?: string | null;
	external_config_reference?: string;
}

export interface RovoAgentCreatorInfo {
	type: 'CUSTOMER' | 'SYSTEM' | 'THIRD_PARTY' | 'FORGE';
	name?: string;
	profileLink?: string;
	id?: string;
}

export interface RovoAgentProfileCardInfo extends RovoAgent {
	creatorInfo: RovoAgentCreatorInfo | undefined;
}
export interface ProfileCardClientData {
	isBot: boolean;
	/**
	 * Mark it as optional as part of the migration from cloud user to AGG user
	 * This field will be calculated based on the user id and the principal user id props
	 * This will be removed once the migration is complete
	 */
	isCurrentUser?: boolean;
	avatarUrl?: string;
	email?: string;
	fullName?: string;
	location?: string;
	meta?: string;
	nickname?: string;
	companyName?: string;
	timestring?: string;
	status: StatusType;
	statusModifiedDate?: number | null;
	customLozenges?: LozengeProps[];
	accountType?: string;
	isAgent?: boolean;
}

export interface ReportingLinesUserPII {
	name: string;
	picture?: string;
}

export interface ReportingLinesUser {
	accountIdentifier: string;
	identifierType: 'ATLASSIAN_ID' | 'BASE64_HASH' | 'UNKNOWN';
	pii?: ReportingLinesUserPII;
}

export interface TeamCentralReportingLinesData {
	managers?: ReportingLinesUser[];
	reports?: ReportingLinesUser[];
}

export interface ProfileCardResourcedProps {
	userId: string;
	cloudId: string;
	resourceClient: ProfileClient;
	actions?: ProfileCardAction[];
	reportingLinesProfileUrl?: string;
	onReportingLinesClick?: (user: ReportingLinesUser) => void;
	position?: ProfilecardTriggerPosition;
	trigger?: TriggerType;
	children?: React.ReactNode;
	addFlag?: (flag: any) => void;
}

export interface ProfileCardResourcedState {
	visible?: boolean;
	isLoading?: boolean;
	hasError: boolean;
	error?: ProfileCardErrorType;
	data: ProfileCardClientData | null;
	reportingLinesData?: TeamCentralReportingLinesData;
	isKudosEnabled?: boolean;
	kudosDrawerOpen: boolean;
	teamCentralBaseUrl?: string;
}

export interface ProfileCardTriggerProps {
	userId: string;
	/**
    A cloudId can be provided, and we'll verify that the target userId is an
    actual user in the specified site.

    Instead you can omit the cloudId and we won't do such a check.

    If you have a cloudId and only want to show users who are in that site
    then please provide it. If you're a site-less product or don't care about
    verifying that the shown user is in a particular site, don't provide a
    cloudId.
   */
	cloudId?: string;
	autoFocus?: boolean;
	resourceClient: ProfileClient;
	actions?: ProfileCardAction[];
	reportingLinesProfileUrl?: string;
	onReportingLinesClick?: (user: ReportingLinesUser) => void;
	position?: ProfilecardTriggerPosition;
	trigger?: TriggerType;
	children: React.ReactNode;
	testId?: string;
	addFlag?: (flag: any) => void;
	ariaLabel?: string;
	ariaLabelledBy?: string;
	prepopulatedData?: PrepopulatedData;
	disabledAriaAttributes?: boolean;
	onVisibilityChange?: (isVisible: boolean) => void;
	isVisible?: boolean;
	offset?: [number, number];
	product?: string;
	viewingUserId?: string;
	agentActions?: AgentActionsType;
	ariaHideProfileTrigger?: boolean;
}

export interface ProfileCardTriggerState {
	visible?: boolean;
	isLoading?: boolean;
	hasError: boolean;
	error?: ProfileCardErrorType;
	data: ProfileCardClientData | null;
	reportingLinesData?: TeamCentralReportingLinesData;
	shouldShowGiveKudos?: boolean;
	teamCentralBaseUrl?: string;
	kudosDrawerOpen: boolean;
}

export interface TeamProfileCardTriggerState {
	visible?: boolean;
	isLoading?: boolean;
	hasError: boolean;
	error?: any;
	data: Team | null;
	renderError?: boolean;
	shouldShowGiveKudos?: boolean;
	teamCentralBaseUrl?: string;
	kudosDrawerOpen: boolean;
}

export interface TeamProfilecardCoreProps {
	/**
    The id of the user viewing the profile card.

    This is used to determine whether to say that the member count is
    "including you" or not.
   */
	viewingUserId?: string;
	/**
    A list of extra buttons to be displayed at the bottom of the card.
    View Profile is always included by default.
   */
	actions?: ProfileCardAction[];
	/**
    A function allowing products to provide an href for the user avatars in the
    profilecard, e.g. so they can link to user's profile pages.
   */
	generateUserLink?: (userId: string) => string;
	/**
    A function allowing products to provide an onClick handler for when the
    user clicks on a user's avatar or avatar group item.
   */
	onUserClick?: (userId: string, event: React.MouseEvent<Element>) => void;
	/**
    This should be a link to the team's profile page. This will be used for:

    - Wrapping the trigger in a link to the team profile page (unless
      triggerLinkType is `none`).

    - Providing the link for the View Profile action button on the card.
   */
	viewProfileLink: string;
	/**
    An onClick action that navigates to the team's profile page. Something you
    may want, e.g. for an SPA site or tracking analytics of navigation. This
    is optional, just the viewProfileLink will suffice. Will be used for:

    - Adding an onClick to the trigger if the triggerLinkType is
      `clickable-link`.

    - Providing an onClick for the View Profile action button on the card.
   */
	viewProfileOnClick?: (event?: React.MouseEvent<Element>) => void;
}

export interface TeamProfileCardTriggerProps extends TeamProfilecardCoreProps {
	/** The id of the team. */
	teamId: string;
	/**
    The id of the organization that the team belongs to.
    Currently this is unused, but will become necessary in the future.
   */
	orgId: string;
	/** An instance of ProfileClient. */
	resourceClient: ProfileClient;
	/**
    The position relative to the trigger that the card should be displayed in.
   */
	position?: ProfilecardTriggerPosition;
	/**
    The interaction method used to trigger the team profile card to appear.

    - Click is generally recommended, but your needs may vary.

    - Hover works for mouse users, but does not support those who use a
      keyboard or screen reader, avoid using this if it's possible or makes
      sense.

    - Hover-click is usable for scenarios like inline-edits, where mouse users
      cannot click on the trigger without causing side effects, but keyboard
      users are still able to navigate into and trigger the profile card.

    Look at the "Team Profilecard Trigger" or "Trigger Link Types" examples to
    see how they behave, or ask in #help-people-and-teams-xpc on Slack for our
    recommendations.
   */
	trigger?: 'hover' | 'click' | 'hover-click';
	/**
    We generally prefer to wrap the trigger in a link to the team profile
    page. This prop determines how that link behaves.

    - Link is generally the recommended prop (especially in combination with
      click or hover-click for the trigger prop above). It wraps the trigger in
      an anchor tag with the team profile link (that users can interact with
      via middle-click, etc.), but left clicking on the link is suppressed.

    - None does not wrap the trigger in a link at all. This makes it difficult
      for keyboard or screen reader users to know how to trigger the profile
      card. Generally avoid this.

    - Clickable-link wraps the trigger in a link with no special behaviour.
      This is suitable for places where you want the trigger to serve primarily
      as a link, and optionally allow hovering to preview the team first.

    Look at the example on "Trigger Link Types" for more in-depth analysis, or
    ask in #help-people-and-teams-xpc on Slack for our recommendations.
   */
	triggerLinkType?: 'none' | 'link' | 'clickable-link';
	/**
    This is the component that will cause a team profile card to appear when
    interacted with according to the method specified by the trigger prop.
   */
	children?: React.ReactNode;
	/**
	 * Used by the card to show Flags.
	 */
	addFlag?: (flag: any) => void;
	/**
	 * Optional cloudId. Pass this if rendering card within a sited context.
	 */
	cloudId?: string;
}

export interface AgentActionsType {
	onChatClick?: () => void;
	onConversationStartersClick?: (starter: ConversationStarter) => void;
}
export interface AgentProfileCardTriggerProps extends AgentActionsType {
	agentId: string;
	cloudId?: string;
	autoFocus?: boolean;
	resourceClient: ProfileClient;
	actions?: ProfileCardAction[];
	position?: ProfilecardTriggerPosition;
	trigger?: TriggerType;
	children: React.ReactNode;
	testId?: string;
	addFlag?: (flag: Flag) => void;
	ariaLabel?: string;
	ariaLabelledBy?: string;
	prepopulatedData?: PrepopulatedData;
	disabledAriaAttributes?: boolean;
	isVisible?: boolean;
	offset?: [number, number];
	product?: string;
	viewingUserId?: string;
}

export type AgentProfileCardProps = {
	resourceClient: ProfileClient;
	agent?: RovoAgentProfileCardInfo;
	isLoading?: boolean;
	hasError?: boolean;
	isCreatedByViewingUser?: boolean;
	cloudId?: string;
	product?: string;
	errorType?: ProfileCardErrorType;
	addFlag?: (flag: Flag) => void;
} & AgentActionsType;

export type StatusType = 'active' | 'inactive' | 'closed';

export type TriggerType = 'hover' | 'click';

export type ProfileType = 'user' | 'team' | 'agent';

export type StatusModifiedDateType =
	| 'noDate'
	| 'thisWeek'
	| 'thisMonth'
	| 'lastMonth'
	| 'aFewMonths'
	| 'severalMonths'
	| 'moreThanAYear';

export interface ProfileCardAction {
	callback?: (...args: any[]) => any;
	shouldRender?: (data: any) => boolean;
	id?: string;
	label: React.ReactNode;
	// Link to provide general link behaviour to the button. If both link and callback are provided link behaviour will be suppressed on click.
	link?: string;
}

export type LozengeColor = 'default' | 'success' | 'removed' | 'inprogress' | 'new' | 'moved';

export interface LozengeProps {
	text: React.ReactNode;
	appearance?: LozengeColor; // defaults to 'default'
	isBold?: boolean; // defaults to false
}

export interface ProfilecardProps {
	isLoading?: boolean;
	hasError?: boolean;
	errorType?: ProfileCardErrorType;
	accountType?: string;
	status?: StatusType;
	isBot?: boolean;
	avatarUrl?: string;
	fullName?: string;
	meta?: string;
	userId?: string;
	isCurrentUser?: boolean;
	// Nick name is also known as public name
	nickname?: string;
	email?: string;
	location?: string;
	companyName?: string;
	timestring?: string;
	actions?: ProfileCardAction[];
	clientFetchProfile?: () => void;
	statusModifiedDate?: number | null;
	withoutElevation?: boolean;
	/** Show manager and direct reports section on profile hover card, if available */
	reportingLines?: TeamCentralReportingLinesData;
	/** Base URL to populate href value for manager's and direct reports' user avatar  */
	reportingLinesProfileUrl?: string;
	/** Click handler when user clicks on manager's and direct reports' user avatar, un-clickable otherwise */
	onReportingLinesClick?: (user: ReportingLinesUser) => void;
	isKudosEnabled?: boolean;
	teamCentralBaseUrl?: string;
	addFlag?: (flag: any) => void;
	cloudId?: string;

	// Allow to pass custom message for disabled account which `status` prop is `inactive` or `closed`.
	// `disabledAccountMessage` should not contain react-intl-next components, ex: `FormattedMessage`,
	// because ProfileCard component is wrapped in its own `IntlProvider` and `FormattedMessage` will loads messages of `@atlaskit/profilecard`,
	// not from the consumer of `@atlaskit/profilecard`.
	disabledAccountMessage?: React.ReactNode;
	// Allow to show a status lozenge for disabled account which `status` prop is `inactive` or `closed`
	hasDisabledAccountLozenge?: boolean;
	// Allow consumers to pass in custom lozenges that will be displayed under the heading
	customLozenges?: LozengeProps[];
	openKudosDrawer?: () => void;
	isTriggeredUsingKeyboard?: boolean;
	disabledAriaAttributes?: boolean;
	//overriding agent actions
	agentActions?: AgentActionsType;
}

export type AnalyticsFromDuration = (duration: number) => AnalyticsEventPayload;

export type AnalyticsFunction = (generator: AnalyticsFromDuration) => void;

export interface AnalyticsProps {
	createAnalyticsEvent?: CreateUIAnalyticsEvent;
}

export interface AnalyticsWithDurationProps {
	fireAnalyticsWithDuration: AnalyticsFunction;
}

export interface TeamProfilecardProps extends TeamProfilecardCoreProps {
	/** Indicates whether the team's details are still loading. */
	isLoading?: boolean;
	/** Indicates whether an error occurred whilst fetching team details. */
	hasError?: boolean;
	/** Describes the type of error that occurred, if any. */
	errorType?: TeamProfileCardErrorType;
	/** The details of the team to be shown. */
	team?: Team;
	/** A callback that will try to re-fetch data in case an error occurred. */
	clientFetchProfile?: () => void;
	/** Details relevant to passing around analytics. */
	analytics: AnalyticsFunction;
}

export interface MessageIntlProviderProps {
	children: React.ReactNode;
	intl: IntlShape;
}

export type RelativeDateKeyType =
	| 'ThisWeek'
	| 'ThisMonth'
	| 'LastMonth'
	| 'AFewMonths'
	| 'SeveralMonths'
	| 'MoreThanAYear'
	| null;

export type AgentIdType = { type: 'agent' | 'identity'; value: string };

type AgentPermissionName = 'AGENT_UPDATE' | 'AGENT_DEACTIVATE';
export type AgentPermissions = {
	permissions: Record<
		AgentPermissionName,
		{
			permitted: boolean;
		}
	>;
};

export interface ProfileClient {
	flushCache: () => void;
	getProfile: (
		cloudId: string,
		userId: string,
		analytics?: (event: AnalyticsEventPayload) => void,
	) => Promise<ProfileCardClientData>;
	getTeamProfile: (
		teamId: string,
		orgId?: string,
		fireAnalytics?: (event: AnalyticsEventPayload) => void,
	) => Promise<Team>;
	getReportingLines: (userId: string) => Promise<TeamCentralReportingLinesData>;
	shouldShowGiveKudos: () => Promise<boolean>;
	getTeamCentralBaseUrl: (teamCentralScopes?: TeamCentralScopes) => Promise<string | undefined>;
	getRovoAgentProfile: (
		id: AgentIdType,
		fireAnalytics?: (event: AnalyticsEventPayload) => void,
	) => Promise<RovoAgent>;
	getRovoAgentPermissions: (
		id: string,
		fireAnalytics?: (event: AnalyticsEventPayload) => void,
	) => Promise<AgentPermissions>;
	deleteAgent: (
		id: string,
		fireAnalytics?: (event: AnalyticsEventPayload) => void,
	) => Promise<void>;
	setFavouriteAgent: (
		id: string,
		isFavourite: boolean,
		fireAnalytics?: (event: AnalyticsEventPayload) => void,
	) => Promise<void>;
}

export type ProfilecardTriggerPosition =
	| 'bottom-start'
	| 'bottom'
	| 'bottom-end'
	| 'left-start'
	| 'left'
	| 'left-end'
	| 'top-end'
	| 'top'
	| 'top-start'
	| 'right-end'
	| 'right'
	| 'right-start';

export type ProfileCardErrorType = {
	reason: 'default' | 'NotFound';
} | null;

export type TeamProfileCardErrorType = {
	reason: 'default' | 'NotFound' | 'TEAMS_FORBIDDEN';
} | null;

export interface ProfileClientOptions
	extends Omit<TeamCentralCardClientOptions, 'gatewayGraphqlUrl'> {
	gatewayGraphqlUrl?: string;
	url?: string;
	cacheSize?: number;
	cacheMaxAge?: number;
	/** Name of integrating product e.g. jira, atlas, confluence **/
	productIdentifier?: string;
	cloudId?: string;
}

export interface ClientOverrides {
	userClient?: UserProfileCardClient;
	teamClient?: TeamProfileCardClient;
	teamCentralClient?: TeamCentralCardClient;
	rovoAgentClient?: RovoAgentCardClient;
}

/** This interface represents the data that is prepopulated in the profile card. **/
export interface PrepopulatedData {
	fullName?: string;
}

export type TeamsUserQueryResponse = {
	id: string;
	name: string;
	picture: string;
	accountStatus: StatusType;
	__typename: 'AppUser' | string;
	email?: string;
	nickname?: string;
	zoneinfo?: string;
	extendedProfile?: {
		jobTitle?: string;
		organization?: string;
		location?: string;
		closedDate?: number;
		inactiveDate?: number;
	};
	appType?: string;
};
