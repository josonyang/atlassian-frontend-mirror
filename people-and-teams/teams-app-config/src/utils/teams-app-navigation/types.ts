export type HostProduct = 'jira' | 'confluence' | 'home';

// Require at least one of orgId or cloudId
type RequireOrgIdOrCloudId =
	| {
			orgId: string;
			/**
			 * Optional, but should always be provided if possible.
			 * Without some or both of the Org and Cloud IDs, there are no guarantees
			 * that the user will land in the teams app in the correct context.
			 */
			cloudId?: string;
	  }
	| {
			/**
			 * Optional, but should always be provided if possible.
			 * Without some or both of the Org and Cloud IDs, there are no guarantees
			 * that the user will land in the teams app in the correct context.
			 */
			orgId?: string;
			cloudId: string;
	  };

export type NavigationActionCommon = RequireOrgIdOrCloudId & {
	/**
	 * @deprecated this will only be used until we remove the embedded directory.
	 * Once traffic is fully migrated to the teams app, this can be removed.
	 */
	push?: (url: string) => void;
	/**
	 * @deprecated this is only required until we remove the embedded directory.
	 * Once traffic is fully migrated to the teams app, this can be removed.
	 */
	hostProduct?: HostProduct;
	/**
	 * By default, the navigation will open in a new tab once traffic is migrated to the teams app.
	 * If the traffic is not migrated, the default will be for the navigation to open in the same tab.
	 */
	shouldOpenInSameTab?: boolean;
	/**
	 * @deprecated this is only required until we remove the embedded directory.
	 * Once traffic is fully migrated to the teams app, this can be removed.
	 * Nav4 is a precondition to the Teams App migration.
	 */
	userHasNav4Enabled?: boolean;
};

export type NavigationActionUser = {
	userId: string;
	// Extend as required
	section?: 'workswith';
};
export type NavigationActionUserWork = {
	userId: string;
};
export type NavigationActionTeam = {
	teamId: string;
};
export type NavigationActionAgent = {
	agentId: string;
};
export type NavigationActionKudos = {
	kudosId: string;
};

export type NavigationActionPeople = {
	query?: string;
};

export type NavigationActionPayloadVariants =
	| {
			type: 'LANDING' | 'DIRECTORY' | 'TEAMS_DIRECTORY';
	  }
	| {
			type: 'USER';
			payload: NavigationActionUser;
	  }
	| {
			type: 'TEAM';
			payload: NavigationActionTeam;
	  }
	| {
			type: 'AGENT';
			payload: NavigationActionAgent;
	  }
	| {
			type: 'KUDOS';
			payload: NavigationActionKudos;
	  }
	| {
			type: 'PEOPLE_DIRECTORY';
			payload: NavigationActionPeople;
	  }
	| {
			type: 'USER_WORK';
			payload: NavigationActionUserWork;
	  };

export type NavigationAction = NavigationActionCommon & NavigationActionPayloadVariants;

export type NavigationResult = {
	onNavigate: (e?: React.MouseEvent | React.KeyboardEvent) => void;
	href: string;
	target: '_blank' | '_self';
};
