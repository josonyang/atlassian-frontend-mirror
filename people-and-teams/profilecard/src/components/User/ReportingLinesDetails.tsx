import React from 'react';

import { FormattedMessage, useIntl } from 'react-intl-next';

import Avatar from '@atlaskit/avatar';
import AvatarGroup, { type AvatarGroupProps } from '@atlaskit/avatar-group';
import { cssMap } from '@atlaskit/css';
import { fg } from '@atlaskit/platform-feature-flags';
import { componentWithFG } from '@atlaskit/platform-feature-flags-react';
// eslint-disable-next-line @atlaskit/design-system/no-emotion-primitives -- to be migrated to @atlaskit/primitives/compiled â€“ go/akcss
import { Box, xcss } from '@atlaskit/primitives';
import { Pressable } from '@atlaskit/primitives/compiled';
import { token } from '@atlaskit/tokens';

import messages from '../../messages';
import {
	ManagerName,
	ManagerSection,
	OffsetWrapper,
	ReportingLinesSection,
} from '../../styled/ReportingLines';
import {
	type AnalyticsWithDurationProps,
	type ProfilecardProps,
	type ReportingLinesUser,
} from '../../types';
import { PACKAGE_META_DATA, reportingLinesClicked } from '../../util/analytics';
import { getPageTime } from '../../util/performance';

import { default as ReportingLinesDetailsCompiled } from './ReportingLinesDetailsCompiled';

export type ReportingLinesDetailsProps = Pick<
	ProfilecardProps,
	'reportingLines' | 'reportingLinesProfileUrl' | 'onReportingLinesClick'
> &
	AnalyticsWithDurationProps;

function getProfileHref(userId: string, profileUrl?: string) {
	return profileUrl ? profileUrl + userId : undefined;
}

const reportingLinesHeadingDefaultStyles = xcss({
	color: 'color.text',
	font: 'font.heading.xxsmall',
	fontWeight: 'font.weight.semibold',
	marginBottom: 'space.100',
});
const reportingLinesHeadingStyles = xcss({
	marginBottom: '0',
});

const styles = cssMap({
	reportingLinesButton: {
		paddingTop: token('space.0'),
		paddingRight: token('space.0'),
		paddingBottom: token('space.0'),
		paddingLeft: token('space.0'),
		backgroundColor: token('color.background.neutral.subtle'),
		'&:hover': {
			backgroundColor: token('color.background.neutral.subtle.hovered'),
		},
	},
});

const avatarGroupMaxCount = 5;

const ReportingLinesDetails = (props: ReportingLinesDetailsProps) => {
	const { formatMessage } = useIntl();
	const {
		fireAnalyticsWithDuration,
		fireAnalyticsWithDurationNext,
		reportingLines = {},
		reportingLinesProfileUrl,
		onReportingLinesClick,
	} = props;
	const { managers = [], reports = [] } = reportingLines;

	const manager = managers.length >= 1 ? managers[0] : undefined;
	const hasReports = reports.length > 0;

	const getReportingLinesOnClick = (
		user: ReportingLinesUser,
		userType: 'manager' | 'direct-report',
	) =>
		onReportingLinesClick
			? () => {
					if (fg('ptc-enable-profile-card-analytics-refactor')) {
						fireAnalyticsWithDurationNext('ui.profilecard.clicked.reportingLines', (duration) => ({
							duration,
							userType,
							firedAt: Math.round(getPageTime()),
							...PACKAGE_META_DATA,
						}));
					} else {
						fireAnalyticsWithDuration((duration) =>
							reportingLinesClicked({
								duration,
								userType,
							}),
						);
					}
					onReportingLinesClick(user);
				}
			: undefined;

	const onReportingLinksClick = (
		user: ReportingLinesUser,
		userType: 'manager' | 'direct-report',
		href: string | undefined,
	) => {
		let shouldPreventDefault = false;
		if (onReportingLinesClick) {
			shouldPreventDefault = onReportingLinesClick(user) === false;
		}

		if (fg('ptc-enable-profile-card-analytics-refactor')) {
			fireAnalyticsWithDurationNext('ui.profilecard.clicked.reportingLines', (duration) => ({
				duration,
				userType,
				firedAt: Math.round(getPageTime()),
				...PACKAGE_META_DATA,
			}));
		} else {
			fireAnalyticsWithDuration((duration) =>
				reportingLinesClicked({
					duration,
					userType,
				}),
			);
		}
		if (shouldPreventDefault) {
			return;
		}

		if (href) {
			window.location.href = href;
		}
	};

	const showMoreButtonProps: AvatarGroupProps['showMoreButtonProps'] = {
		'aria-label': formatMessage(messages.profileCardMoreReportingLinesLabel, {
			count: reports.length - avatarGroupMaxCount + 1,
		}),
	};

	return (
		<>
			{manager && (
				<ReportingLinesSection>
					<Box xcss={[reportingLinesHeadingDefaultStyles, reportingLinesHeadingStyles]}>
						<FormattedMessage {...messages.managerSectionHeading} />
					</Box>
					<OffsetWrapper>
						<Pressable
							onClick={() =>
								onReportingLinksClick(
									manager,
									'manager',
									getProfileHref(manager.accountIdentifier, reportingLinesProfileUrl),
								)
							}
							isDisabled={!onReportingLinesClick}
							xcss={styles.reportingLinesButton}
						>
							<ManagerSection>
								<Avatar size="xsmall" src={manager.pii?.picture} />
								<ManagerName>{manager.pii?.name}</ManagerName>
							</ManagerSection>
						</Pressable>
					</OffsetWrapper>
				</ReportingLinesSection>
			)}
			{hasReports && (
				<ReportingLinesSection>
					<Box xcss={reportingLinesHeadingDefaultStyles}>
						<FormattedMessage {...messages.directReportsSectionHeading} />
					</Box>
					<AvatarGroup
						appearance="stack"
						size="small"
						data={reports.map((member: ReportingLinesUser) => {
							return {
								key: member.accountIdentifier,
								name: member.pii?.name || '',
								src: member.pii?.picture,
								href: getProfileHref(member.accountIdentifier, reportingLinesProfileUrl),
								onClick: getReportingLinesOnClick(member, 'direct-report'),
							};
						})}
						maxCount={avatarGroupMaxCount}
						testId="profilecard-reports-avatar-group"
						showMoreButtonProps={showMoreButtonProps}
					/>
				</ReportingLinesSection>
			)}
		</>
	);
};

const ReportingLinesDetailsExport = componentWithFG(
	'profilecard_primitives_compiled',
	ReportingLinesDetailsCompiled,
	ReportingLinesDetails,
)
export default ReportingLinesDetailsExport;
