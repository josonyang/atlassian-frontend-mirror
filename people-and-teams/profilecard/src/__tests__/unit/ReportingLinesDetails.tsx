import React from 'react';

import { render, screen } from '@testing-library/react';
import { IntlProvider } from 'react-intl-next';

import ReportingLinesDetails, {
	type ReportingLinesDetailsProps,
} from '../../components/User/ReportingLinesDetails';
import { type ReportingLinesUser } from '../../types';

const exampleReportingLineUser: ReportingLinesUser = {
	accountIdentifier: 'abcd',
	identifierType: 'ATLASSIAN_ID',
	pii: {
		name: 'manager name',
		picture: 'picture',
	},
};
const exampleReportingLineUser2: ReportingLinesUser = {
	accountIdentifier: 'abcd2',
	identifierType: 'ATLASSIAN_ID',
	pii: {
		name: 'name2',
		picture: 'picture2',
	},
};
const exampleReportingLineUser3: ReportingLinesUser = {
	accountIdentifier: 'abcd3',
	identifierType: 'ATLASSIAN_ID',
	pii: {
		name: 'name3',
		picture: 'picture3',
	},
};
describe('ReportingLinesDetails', () => {
	const defaultProps: ReportingLinesDetailsProps = {
		reportingLines: {
			managers: [exampleReportingLineUser],
			reports: [exampleReportingLineUser2, exampleReportingLineUser3],
		},
		reportingLinesProfileUrl: 'profile-url',
		fireAnalyticsWithDuration: () => {},
		fireAnalyticsWithDurationNext: () => {},
	};

	const renderComponent = (props = {}) =>
		render(
			<IntlProvider locale="en">
				<ReportingLinesDetails {...defaultProps} {...props} />
			</IntlProvider>,
		);

	it('should render mangaer & reports when specified', () => {
		renderComponent({ onReportingLinesClick: () => null });
		expect(screen.getByText('Manager')).toBeInTheDocument();
		expect(screen.getByText('manager name')).toBeInTheDocument();
		expect(screen.getByText('Direct reports')).toBeInTheDocument();
		// img alts
		expect(screen.getByAltText('name2')).toBeInTheDocument();
		expect(screen.getByAltText('name3')).toBeInTheDocument();
	});

	it('should not render mangaer & reports when NOT specified', () => {
		renderComponent({ reportingLines: {} });
		expect(screen.queryByText('Manager')).not.toBeInTheDocument();
		expect(screen.queryByText('Direct reports')).not.toBeInTheDocument();
	});
});
