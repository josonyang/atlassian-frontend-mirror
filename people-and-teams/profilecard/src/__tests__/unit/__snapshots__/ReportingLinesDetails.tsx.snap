// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportingLinesDetails should match snapshot when manager and reports are NOT specified 1`] = `<Fragment />`;

exports[`ReportingLinesDetails should match snapshot when manager and reports are specified 1`] = `
<Fragment>
  <Styled(div)>
    <Styled(h3)
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      <MemoizedFormattedMessage
        defaultMessage="Manager"
        description="Title for a section on the profile card that show the user's direct manager"
        id="pt.team-profile-card.manager.heading"
      />
    </Styled(h3)>
    <Styled(div)>
      <Button
        appearance="subtle"
        href="profile-urlabcd"
        isDisabled={false}
        onClick={[Function]}
        spacing="none"
      >
        <Styled(div)>
          <Avatar
            size="xsmall"
            src="picture"
          />
          <Styled(span)>
            name
          </Styled(span)>
        </Styled(div)>
      </Button>
    </Styled(div)>
  </Styled(div)>
  <Styled(div)>
    <Styled(h3)>
      <MemoizedFormattedMessage
        defaultMessage="Direct reports"
        description="Title for a section on the profile card that show the user's direct reports"
        id="pt.team-profile-card.directReports.heading"
      />
    </Styled(h3)>
    <AvatarGroup
      appearance="stack"
      data={
        Array [
          Object {
            "href": "profile-urlabcd",
            "key": "abcd",
            "name": "name",
            "onClick": [Function],
            "src": "picture",
          },
          Object {
            "href": "profile-urlabcd",
            "key": "abcd",
            "name": "name",
            "onClick": [Function],
            "src": "picture",
          },
        ]
      }
      maxCount={5}
      size="small"
      testId="profilecard-reports-avatar-group"
    />
  </Styled(div)>
</Fragment>
`;
